Rc/manage user tableau permissions

Robert-Costello opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->
Adds role option for managing users' tableau configuration. This is a sub-permission of "Web Users"
<img width="625" alt="Screenshot 2024-04-19 at 5 20 43 PM" src="https://github.com/dimagi/commcare-hq/assets/42388648/19f9323a-e181-4b9a-a9fc-9299d9468483">

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
[Jira ticket](https://dimagi.atlassian.net/browse/USH-4449)
[tech spec](https://docs.google.com/document/d/1TUUhpHBKFJAs7tES99CKS8xkavYLUgqiX0PwQVcvT4s/edit#heading=h.d7s2sycdfk92)
[related model change PR](https://github.com/dimagi/commcare-hq/pull/34454)

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
[Tableau User Syncing](https://www.commcarehq.org/hq/flags/edit/tableau_user_syncing/)

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
locally tested and tested on staging. The changes are also specific to Tableau User Syncing and as of now, very few domains have that FF enabled. This reduces the blast radius.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
Added tests that without the required permission:
- User bulk download does not include tableau configs
- User can't bulk import tableau configs

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
no QA

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations
This can be reverted without further consideration up to the point where users begin managing tableau edit permissions through this UI. After that point some external communications will be needed, though a revert should not impact functionality outside of this feature.

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

Robert-Costello committed:
add tableau permissions options to edit role ui

Robert-Costello committed:
normalize view permission

Robert-Costello committed:
update context for edit web user template

Robert-Costello committed:
check permissions for view and edit tableau config

Jtang-1 committed:
Merge branch 'master' into rc/manage_user_tableau_permissions

Jtang-1 committed:
fix to view permission if only edit is toggled on

it isn't until permissions are saved that they are normalized so
while the modal is being updated, toggling edit permission to true
does not also toggle view permission to true.

Jtang-1 committed:
show tableau edit permission only if user has permission to edit webuser

data binding needs to handle either a boolean or a ko computed function

Jtang-1 committed:
manage tableau config permissions are subpermissions of web users permissions

Jtang-1 committed:
tableau config permission is a subpermission fo web user

Jtang-1 committed:
create readonly view for tableau form

Jtang-1 committed:
display confirmation to user on successful save

Robert-Costello commented:
@Jtang-1 the changes you added look good to me. I tagged a couple more devs for review since the two of us contributed to the changes here.

Jtang-1 committed:
keep permission behaviors for existing web users

Jtang-1 committed:
provides instructions for manually migration upon failure

if failure occurs from changes to the code (such as model changes)
then linked changelog will provide instructions to properly manually migrate

Jtang-1 commented on corehq/apps/users/static/users/js/roles.js:
```diff
@@ -225,6 +232,26 @@ hqDefine('users/js/roles',[
+                            return self.permissions.edit_web_users();
+                        }),
+                        editCheckboxLabel: "edit-user-tableau-config-checkbox",
+                        showViewCheckbox: true,
```
I guess you can saw the view checkbox is still enabled but it's hidden since the parent container `showOption` is hidden ([here](https://github.com/dimagi/commcare-hq/blob/4b92dc891a0be748bdb36ae97e684724dc2396ee/corehq/apps/users/templates/users/partials/edit_role_modal.html#L47))

Jtang-1 commented on corehq/apps/users/static/users/js/roles.js:
```diff
@@ -225,6 +232,26 @@ hqDefine('users/js/roles',[
+                        showAllowCheckbox: false,
+                        allowCheckboxText: null,
+                        allowCheckboxId: null,
+                        allowCheckboxPermission: null,
```
That's for the additional checkbox below the permissions text so not applicable here ![Screen Shot 2024-05-15 at 10 44 44 PM](https://github.com/dimagi/commcare-hq/assets/88759246/40a624b7-5cf4-4798-aeff-84478a730a44)

Jtang-1 commented on corehq/apps/users/templates/users/partials/edit_role_modal.html:
```diff
@@ -55,7 +55,7 @@ <h4 class="modal-title" data-bind="text: modalTitle"></h4>
                   </div>
                 </div>
-                <div class="col-sm-2 controls" data-bind="visible: !showEditCheckbox">
+                <div class="col-sm-2 controls" data-bind="visible: !ko.utils.unwrapObservable(showEditCheckbox)">
```
yes, and it took me way too long to figure how do that before remembering `ko.utils.unwrapObservable` was a thing

Jtang-1 committed:
backend validation for user permission before saving

Jtang-1 commented on corehq/apps/users/forms.py:
```diff
@@ -1786,6 +1786,7 @@ class TableauUserForm(forms.Form):
     )
 
     def __init__(self, *args, **kwargs):
```
good call 4[4866745](https://github.com/dimagi/commcare-hq/pull/34468/commits/486674533c823296afab763dd7f3de4b80017682)

Jtang-1 commented on migrations.lock:
```diff
@@ -1223,6 +1223,7 @@ users
  0062_add_web_apps_groups_permission_flag
  0063_rename_location_invitation_primary_location_and_more
  0064_add_edit_view_tableau_config_permissions
+ 0065_transfer_to_tableau_config_permission
```
Why do you think they should have defaulted true? I'm thinking defaulting false and doing this migration (as we're doing now) are the right steps.

orangejenny replied:
Just adding another comment in favor of defaulting to false and doing the migration. You can default couch attributes to true, but for permissions we almost certainly don't want to.

Jtang-1 committed:
simplify logic as only very few domains currently have FF enabled

filtering by domains that have the FF enabled is enoough to reduces the number of
roles to consider for migration to a manageable amount

Jtang-1 commented on corehq/apps/users/migrations/0065_transfer_to_tableau_config_permission.py:
```diff
@@ -0,0 +1,60 @@
+
+@skip_on_fresh_install
+def transfer_web_user_permission_to_tableau_config_permission(apps, schema_editor):
+    user_role_ids_to_migrate = get_user_role_ids_to_migrate()
```
That's a fair point, KISS principle [00d7ec6](https://github.com/dimagi/commcare-hq/pull/34468/commits/00d7ec6e958f18c2813e33850e9b881cc621cc80)

Jtang-1 commented on corehq/apps/users/migrations/0065_transfer_to_tableau_config_permission.py:
```diff
@@ -0,0 +1,115 @@
+    try:
+        transfer_web_user_permission_to_tableau_config_permission()
+        migrated = len(get_user_role_ids_to_migrate()) == 0
+        if not migrated:
```
Does this address your comment? [00d7ec6](https://github.com/dimagi/commcare-hq/pull/34468/commits/00d7ec6e958f18c2813e33850e9b881cc621cc80)

Jtang-1 committed:
bug: Ensure Tableau config permission is only granted and option shown  when FF is enabled

Prior to this fix, when the FF was disabled and a role was granted web user edit/view permission, the role was inadvertently given Tableau config permission in the background.

With this change, Tableau config permission is only provided when the FF is enabled and explicitly set. This ensures consistency and prevents unintended permission grants.

Jtang-1 committed:
adds link to changelog

snopoke reviewed (commented):
Considering the Django limitation of accessing custom methods on models during a migration I think it would be much better to provide a mechanism for updating role permissions that is safe to use during a migration rather than going with a 'managed migration' approach.

There are already some migrations which do similar things: https://github.com/dimagi/commcare-hq/blob/92c66917f23593fc1588c88f00ceee494aa8adbc/corehq/apps/users/migrations/0037_add_edit_messaging_permission.py#L11-L17

Jtang-1 committed:
include tableau config in download files only if owner has permission to view tableau config

Jtang-1 committed:
refactor: use historical model instead of imported model

refactor had to be done because historical model does not contain custom methods

Jtang-1 commented:
Thanks for the suggestion Simon, this is much better approach than having to issue a changelog. Refactored to migration [4d5c8de](https://github.com/dimagi/commcare-hq/pull/34468/commits/4d5c8de49e8e395dc89699651161debe95f6596b)

Separate unrelated Comment on the PR:
This PR will now also include restrictions on bulk download and upload for tableau config according to the owner's permissions. Thus the base branch was changed from `master` to the in progress work `rc/add-tableau-role-group-import-columns`

Jtang-1 committed:
restrict editing tableau role and tableau groups to only users with edit tableau config permissions

Jtang-1 committed:
test: pass in permissions filter

Jtang-1 committed:
bug: handle case of where no rows are provided in uploaded file

Jtang-1 committed:
tests bulk import for tableau config with and without edit permission

Jtang-1 commented:
Closing and reopening to run tests

Separate note: base branch was changed back to master as the changes did not depend on rc/add-tableau-role-group-import-columns

Jtang-1 commented on corehq/apps/users/permissions.py:
```diff
@@ -75,3 +75,13 @@ def has_permission_to_view_report(couch_user, domain, report_to_check):
     )
+
+
+def get_user_has_permissions_dict(couch_user, domain: str):
```
Ya, the only thing is that it doesn't account for if user is an admin which `has_permission` does ([ref](https://github.com/dimagi/commcare-hq/blob/3a44014879131d0b87432bdde1a3d7fa028ead5b/corehq/apps/users/models.py#L442))

Jtang-1 replied:
I think I get what you're saying, something like this?
```
status = {}
dm = web_user.get_domain_membership(...)
permissions = dm.role.permissions
for p_name in HqPermissions.permission_names():
    status[p_name] = dm.role.permissions[p_name]
```
But there's these admin [checks within "has_permission"](https://github.com/dimagi/commcare-hq/blob/2696ffadba5c14b1adf73cb252b878daf08abeb6/corehq/apps/users/models.py#L639-L642) that grants the user permission even if the assigned role does not have permission. I don't want to copy all that logic into `get_user_has_permissions_dict`.

Taking a step back, the reason I decided to do it this way is to avoid passing a WebUser object as an argument for `parse_web_users` to reduce coupling. Maybe that was the wrong decision. Now i'm leaning towards passing in the WebUser and doing the check owner.has_permission(domain, "edit_view_tableu_config"), what do you think?

Jtang-1 replied:
sounds good, made that change here [0b6983a](https://github.com/dimagi/commcare-hq/pull/34468/commits/0b6983a0b1b142c973632bc81f192c53d4b97aa1)

Jtang-1 commented on corehq/apps/users/migrations/0065_transfer_to_tableau_config_permission.py:
```diff
@@ -0,0 +1,73 @@
+            has_view_web_users_permission = role.rolepermission_set.filter(permission_fk=view_web_users_permission).exists()
+
+            if has_edit_web_users_permission:
+                role.rolepermission_set.get_or_create(permission_fk=edit_user_tableau_config_permission, defaults={"allow_all": True})
```
It doesn't explicitly set anything to true. It creates the `RolePermission` for edit/view tableau config and assigns the `UserRole` to the `role` `ForeignKey` field. `role.rolepermission_set` contains all the permissions that role has, which will now include edit/view tableau config.

For example,
```
In [26]: for rolepermission in user.rolepermission_set.all():
    ...:     print(rolepermission.permission)
    ...: report_an_issue
access_mobile_endpoints
view_groups
edit_commcare_users
view_commcare_users
edit_locations
view_locations
access_all_locations
access_api
edit_groups
edit_shared_exports
view_web_users
edit_user_tableau_config
view_user_tableau_config
edit_web_users
```

Jtang-1 committed:
refactor: directly pass web user to use for checking permission

The previous strategy was to decouple the function from an Owner but
resulted in overcomplication and wrapping HQPermission multiple times.

Jtang-1 commented on corehq/apps/users/migrations/0065_transfer_to_tableau_config_permission.py:
```diff
@@ -0,0 +1,73 @@
+
+    def transfer_web_user_permission_to_tableau_config_permission():
+        status = get_migration_status(ALL_DOMAINS, MIGRATION_SLUG)
+        if status == MigrationStatus.COMPLETE:
```
I think it would still be helpful for logging purposes in case the migration fails for some domains. What do you mean by "migration is restartable"?  Do you mean if the migration logic was in a management command that could be rerun?

Jtang-1 replied:
If there is an exception, some domains would have already been migrated and others would not have (I'm pretty sure that's right?). In that situation, tracking the status would record in `DomainMigrationProgress` model which domains have and haven't been completed yet. This would be more fidelity than just the overall Django migration status. 

I don't think it matters much here since there's only a handful of domains. But I think that's what the additional purpose would be.

Jtang-1 commented on corehq/apps/users/migrations/0065_transfer_to_tableau_config_permission.py:
```diff
@@ -0,0 +1,73 @@
+    except Exception:
+        traceback.print_exc()
+        print("")
+        print("Auto migrate failed")
```
Not currently I don't see why it would fail when run automatically but succeed when run manually. I included this previously from I was importing the models in directly which potentially would need manual intervention. I decided to keep it in as a safeguard for any unforeseen problems.
